cmd/main.go

```diff
@@ -328,20 +337,63 @@ func run(opts *options) error {
 		DatadogGenericResourceEnabled: opts.datadogGenericResourceEnabled,
 	}
 
-	if err = controller.SetupControllers(setupLog, mgr, options); err != nil {
+	// Never use original mgr.GetConfig(), always copy as clients might modify the configuration
+	discoveryConfig := rest.CopyConfig(mgr.GetConfig())
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(discoveryConfig)
+	if err != nil {
+		return fmt.Errorf("unable to get discovery client: %w", err)
+	}
+
+	versionInfo, err := discoveryClient.ServerVersion()
+	if err != nil {
+		return fmt.Errorf("unable to get APIServer version: %w", err)
+	}
+
+	if versionInfo != nil {
+		gitVersion := versionInfo.GitVersion
+		if !utils.IsAboveMinVersion(gitVersion, "1.16-0") {
+			setupLog.Error(nil, "Detected Kubernetes version <1.16 which requires CRD version apiextensions.k8s.io/v1beta1. "+
+				"CRDs of this version were removed in v1.10.0.")
+		}
+	}
+
+	groups, resources, err := getServerGroupsAndResources(setupLog, discoveryClient)
+	if err != nil {
+		return fmt.Errorf("unable to get API resource versions: %w", err)
+	}
+	platformInfo := kubernetes.NewPlatformInfo(versionInfo, groups, resources)
+
+	if err = controller.SetupControllers(setupLog, mgr, platformInfo, options); err != nil {
 		return setupErrorf(setupLog, err, "Unable to start controllers")
 	}
 
+	mdf := setupMetadataForwarder(metadataLog, mgr.GetClient(), versionInfo.String(), opts)
+	mdf.Start()
+
 	// +kubebuilder:scaffold:builder
 
 	setupLog.Info("starting manager")
 	if err := mgr.Start(ctrl.SetupSignalHandler()); err != nil {
 		return setupErrorf(setupLog, err, "Problem running manager")
 	}
 
+	// Stop metadata forwarder
+	mdf.Stop()
```

Comment:
Did we move the above block to be able to have `mdf` initialized and stopped in main? version info handling seems a bit low level stuff for main, I'm also wondering if we really need to call Stop() here.

Reply:
> but it should be conscious decision

Yes, it is consciously decided on my part. As for submitting one last payload, I don't think it's valuable because the interval is already at 1 minute.

> does it make sense to move that large block elsewhere to have less clutter in main?

I'll work on factoring that out